Add SIMD Array-Array set operations

saik0 opened the pull request:
Initial vectorized set op implementations.

One noteworthy difference from experimental branch: This does not include galloping array intersection, so may be noticeably slower than previous benchmarks.

Kerollmops reviewed (commented):
You have done awesome work here, and the documentation too!
I was wondering when you are enabling the `portable_simd` unstable feature for when the `simd` crate feature is enabled?

saik0 commented on src/bitmap/store/array_store/mod.rs:
```diff
@@ -251,8 +252,11 @@ impl BitOr<Self> for &ArrayStore {
 
     fn bitor(self, rhs: Self) -> Self::Output {
         #[allow(clippy::suspicious_arithmetic_impl)]
-        let capacity = (self.vec.len() + rhs.vec.len()).min(4096);
+        let capacity = self.vec.len() + rhs.vec.len();
```
Good question! No. I removed it.

Rationale: If the op exceeds the 4096 array limit, this alg continues until completion. It's handled later and converted to a bitmap in `ensure_correct_store`. 

Example, lhs = 0..4096, rhs = 4096..8192. The capacity will need to grow to 8192.
IMO, Better to over-allocate at the beginning than need to realloc mid processing, and then alloc AGAIN for bitmap

saik0 commented on src/bitmap/store/array_store/vector.rs:
```diff
@@ -0,0 +1,766 @@
+
+// a one-pass SSE union algorithm
+pub fn or(lhs: &[u16], rhs: &[u16], visitor: &mut impl BinaryOperationVisitor) {
+    // De-duplicates `slice` in place
+    // Returns the end index of the deduplicated slice.
+    // elements after the return value are not guaranteed to be unique or in order
+    #[inline]
+    fn dedup(slice: &mut [u16]) -> usize {
+        let mut pos: usize = 1;
+        for i in 1..slice.len() {
+            if slice[i] != slice[i - 1] {
+                slice[pos] = slice[i];
+                pos += 1;
+            }
+        }
+        pos
+    }
```
Yeah, though in this specific instance case we can backshift as we do here, instead of swapping to the back.  We don't care about the back half of the binary partition.

saik0 replied:
Besides! This whole module is nightly only due to core simd.

saik0 commented:
> You have done awesome work here, and the documentation too!

Thank you!

> I was wondering when you are enabling the `portable_simd` unstable feature for when the `simd` crate feature is enabled?

Do you mean [`#![cfg_attr(feature = "simd", feature(portable_simd))]`](https://github.com/RoaringBitmap/roaring-rs/blob/master/src/lib.rs#L10). We added that to `lib.rs` in #152

Kerollmops reviewed (approved):
> This does not include galloping array intersection

Is it because you haven't implemented it with `std::simd` (yet), or is it because it can't be done without introducing raw SIMD calls? I don't remember, sorry 😰

Anyway, this PR is wonderful, so, merging it RN 🎉
bors merge

Kerollmops commented:
Failed due to _another_ network error, so retrying.
bors merge

Kerollmops commented:
I need to set up a release one day but you'll tell me when. I am not sure if we broke some API but that would not surprise me if we did. Maybe I'll [use semverver](https://github.com/rust-lang/rust-semverver) to be sure or simply bump the major number.

saik0 commented:
>  I am not sure if we broke some API but that would not surprise me if we did

I don't think we've changed any API, other than adding `rank`.

It feels weird to me to release with `rank` but not `select`

I have select implemented in a local branch, its pretty simple. I implemented it together with `rank`. Only reason I haven't created a PR is because I need to add tests. If I pushed it up do you think you could write some? I'm prioritizing #145

Kerollmops commented:
> I don't think we've changed any API, other than adding `rank`.

Ok, I will check by hand when we are ready to release a new version!

> It feels weird to me to release with `rank` but not `select`
>
> I have select implemented in a local branch, its pretty simple. I implemented it together with `rank`. Only reason I haven't created a PR is because I need to add tests. If I pushed it up do you think you could write some? I'm prioritizing #145

I'm not sure I would be able to implement that myself, as I am not as much at ease with the codebase. There is no rush at all, we can wait as long as we want, tell me when you think we can release something. I will make sure to post it on Reddit because I am proud of what you've made 🎉

However, if `select` is as simple as it seems: `bitmap & (1 << (n - 1))` maybe I could do that. What do you think?
